### ISSUE TITLE
spGEMM produces crashes and wrong results

### ISSUE THREAD

**Issue by dabeschte** [kokkos/kokkos-kernels#402]
for many matrices, the kokkos-kernels spGEMM crashes or when it does not, it often returns wrong results. if it produces wrong results, the outputs matrix C is only 8 bytes large and from investigation in the verbose logs, it also seems like the amount of non zeroes in C does not match with the amount produced by other spGEMM solvers.

one example matrix with an incorrect result and no output matrix, but where it does not crash: [130bit](https://sparse.tamu.edu/Priebel/130bit)

one example matrix which is very small, but it still crashes: [aa01](https://sparse.tamu.edu/Meszaros/aa01)

i am using a Titan V with compute capability 7.0 on Ubuntu 18.04, but according to the github repo and the building process, this should be supported

**Comment by srajama1** [comment]
@dabeschte : Can you post your configuration ? What are you multiplying these matrices with ? Are you using our example driver or are you calling the code from your libraries ?

**Comment by dabeschte** [comment]
@srajama1 Thank you for the fast reply.
I am using you example executable in the perf_test and the matrix is multiplied with itself (I just use --amtx, but not --bmtx).
I use the following parameters for multiplying the matrix:
`./KokkosSparse_spgemm.exe --cuda 0 --amtx 130bit.mtx`

I just had the idea that the crashes could happen, because I use Cuda 10.1, but since the changes from 10.0 to 10.1 seem minor, I think this should not be the issue here.

I used the following build settings:
```
KOKKOS_PATH=${HOME}/repos/kokkos #path to kokkos source
KOKKOSKERNELS_SCALARS=double #the scalar types to instantiate =double,float...
KOKKOSKERNELS_LAYOUTS=LayoutLeft #the layout types to instantiate.
KOKKOSKERNELS_ORDINALS=int #ordinal types to instantiate
KOKKOSKERNELS_OFFSETS=int #offset types to instantiate
KOKKOSKERNELS_PATH=../.. #path to kokkos-kernels top directory.
CXX=${KOKKOS_PATH}/bin/nvcc_wrapper #icpc #
KOKKOSKERNELS_OPTIONS=eti-only #options for kokkoskernels  
KOKKOS_DEVICES=OpenMP,Cuda # other devices Cuda,Serial ..
KOKKOS_ARCHS=KNL
CXXFLAGS="-Wall -pedantic -Werror -O3 -g -Wshadow -Wsign-compare -Wtype-limits -Wuninitialized"
```

**Comment by srajama1** [comment]
@dabeschte This could explain it. I have heard mixed feedback on 10.0 and been asked to wait for 10.1. We do not have continuous builds on 10.1 yet. @ndellingwood : Is this something we could add in kokkos-dev ?

**Comment by dabeschte** [comment]
@srajama1 ok. I could try downgrading to 10.0 and see if that works.
Or do you maybe have a running installation on you system with which you could verify that those matrices should work with other drivers?
I can get a lot of matrices working with my current configuration, yet about half of them have before mentioned issues.

**Comment by srajama1** [comment]
@dabeschte I should have been more clear. 10.1 should clearly be an improvement over 10.0, we just haven't tested it. We were not going to mess with 10.0 due to recommendation from our system software team. All our current builds are on 9.2.

**Comment by dabeschte** [comment]
@srajama1 ok good to know. But this still leaves the question if the bug is due to driver incompatibility or the spgemm implementation. If you could verify whether those matrices work on your configuration, I could avoid a driver downgrade on my system. That would be great!

**Comment by srajama1** [comment]
@dabeschte : Will do and update with my drivers.

**Comment by dabeschte** [comment]
@srajama1 great. Thank you!

**Comment by ndellingwood** [comment]
> KOKKOS_ARCHS=KNL

@dabeschte the arch variable in your build setting is incorrect, to support the sm_70 GPU this should be:
`KOKKOS_ARCHS="Volta70"`

Can you try updating that and let us know if that resolves the issue?

**Comment by ndellingwood** [comment]
> Is this something we could add in kokkos-dev ?

@srajama1 only cuda 8.0 is available on kokkos-dev; we may be able to do this on kokkos-dev-2 but I'm not sure if it is set up or intended to be used for nightlies, we need to check with @crtrott

**Comment by ndellingwood** [comment]
@dabeschte for reference, here is a link for variable names for various architectures you may be interested in compiling for: https://github.com/kokkos/kokkos/wiki/Compiling#table-43-architecture-variables

**Comment by dabeschte** [comment]
@ndellingwood ah, sorry about that. I copied the example build configuration which I tried to use first.
But i looked into my bash history and I definitely used the generate_makefile.bash script with the following params

`./generate_makefile.bash --kokkoskernel-path=/home/mathias/repos/kokkos-kernels --with-scalars=double --kokkos-path=/home/mathias/repos/kokkos --with-cuda=/usr/local/cuda --with-openmp --arch=Volta70 --compiler=/home/mathias/repos/kokkos/bin/nvcc_wrapper`

is this configuration correct?

**Comment by ndellingwood** [comment]
@dabeschte that looks like it should be fine, if it isn't too much trouble could you try modifying the call to `generate_makefile.bash` slightly with a clean build? It likely won't make a difference, but worth trying in case explicitly passing the path to the cuda compiler may be throwing off something.

`./generate_makefile.bash --kokkoskernel-path=/home/mathias/repos/kokkos-kernels --with-scalars=double --kokkos-path=/home/mathias/repos/kokkos --with-device="Cuda,OpenMP" --arch=Volta70 --compiler=/home/mathias/repos/kokkos/bin/nvcc_wrapper`

Also, which branches of kokkos and kokkos-kernels are you using? I'd suggest also testing with the most recent master branches of each (this will make sure they are compatible versions that we have tested well), in addition to making sure the develop branches of each are updated and in sync if you are testing those branches.

**Comment by dabeschte** [comment]
@ndellingwood sure, I will try that at monday when I'm back in office.
But I also doubt that it will make a difference since the building process could complete and some matrices seem to work.

**Comment by dabeschte** [comment]
@ndellingwood unfortunately, using your build parameters did not help.
I am currently rebuilding kokkos-kernels without OpenMP to see if this makes a difference. Will update when I have new information.

**Comment by srajama1** [comment]
@dabeschte : I should have caught this much sooner. You are using the test driver option for A*A with a rectangular matrix which is causing this. We do have an option for A*B to which you can pass A' as the B. I just transposed and saved aa01 and used that option. That works.

**Comment by srajama1** [comment]
We will modify the driver to catch this and return a reasonable error.

**Comment by dabeschte** [comment]
@srajama1 that makes sense and that was also my first thought, but since [adder_dcop_01](https://sparse.tamu.edu/Sandia/adder_dcop_01) is a square matrix and it also return incorrect results, my guess was that there is another issue in the code. other examples which either crash or produce wrong results are [baumann](https://sparse.tamu.edu/Watson/Baumann) and [bauru5727](https://sparse.tamu.edu/Rommes/bauru5727) which are reasonable large for a Titan V and are both square.

**Comment by srajama1** [comment]
@dabeschte I will try these square matrices and let you know.

**Comment by dabeschte** [comment]
@srajama1 any news about those matrices?

**Comment by delanyinspiron6400** [comment]
I'm having the same problems with the TITAN Xp on CUDA 9.2 and 10.0/10.1, tried to test as well on the SuiteSparse matrix collection and a few hundred seem to not work. 
This includes for example: [3elt](https://sparse.tamu.edu/AG-Monien/3elt) , [add20](https://sparse.tamu.edu/Hamm/add20) and [192bit](https://sparse.tamu.edu/Priebel/192bit).
The same is true for those mentioned by @dabeschte, any news here?

**Comment by srajama1** [comment]
Sorry, got pulled into other things. I will try these ...

**Comment by srajama1** [comment]
@ndellingwood : Can you help me reproduce this ? If this is a problem, it would be nice to fix before 2.9.

**Comment by srajama1** [comment]
@ndellingwood : Ping

**Comment by ndellingwood** [comment]
I spent a bit of time digging into this today. Cases like the `aa01` matrix were problematic when run with the spgemm perf test driver code `KokkosSparse_spgemm.exe` because that code will internally use the `Impl::read_kokkos_crst_matrix` routine to read the matrices which is does not properly handle matrix market format files with the `pattern` field (for example adjacency matrices without a "values" field). 
I used Matlab to regenerate that matrix market file with the `real` field and 3rd field of values, created the transpose matrix and also write it matrix market file, then ran the spgemm code with these matrices and the spgemm results matched those from matlab.

I have not ruled out yet if problems with the other matrices (I tried `adder_dcop_01` and had discrepancy in results with Matlab) are due to file IO issues with the internal matrix reader utilities or in the spgemm code, this will take a bit more time to investigate and I don't have much time in the near term to dig beyond ruling out file IO as an issue.

**Comment by ndellingwood** [comment]
I believe the perf_test `KokkosSparse_spgemm` code was not written for general testing but rather performance testing within a specific scripted framework, so like mentioned before at least sanity checks should be added that error out with useful messages (matrix dimensions not compatible for multiplication, matrix market field not supported, etc).

**Comment by srajama1** [comment]
@ndellingwood : I can help in getting the read_kokkos_crs_matrix fixed. Let me start on that. Thanks for narrowing it down, I don't know why these problems don't show up in other places.

**Comment by brian-kelley** [comment]
@srajama1 @dabeschte The fix for this is in develop now. I tested it with adder_dcop_01.mtx, aa01.mtx and 130bit.mtx and SPGEMM_KKMEM produces the correct result on all of them (by comparing with matlab).

The remaining issue is reading "pattern" MatrixMarket files, which I'm working on now. When I did the testing above, I had to set all the scalar values to 1 in matlab to get a regular "coordinate" sparse matrix.

**Comment by dabeschte** [comment]
ok cool. was the issue with adder_dcop_01.mtx and aa01.mtx matrices also because of the mtx-file reader or was this a different bug?

**Comment by brian-kelley** [comment]
@dabeschte adder_dcop_01 is a regular coordinate matrix so the mtx reader worked fine on it. The bug there was the actual SPGEMM algorithm bug that just got fixed. aa01 and 130bit are both "pattern" matrices (really graphs) and were only ever affected by the mtx reader issue.

Out of curiosity, I tried running version of SPGEMM before today's fix on aa01 and 130bit with the workaround of setting each scalar value to 1.0, and SPGEMM was correct for both of them.
